### PR TITLE
feat: added retry logic while estimating fee

### PR DESCRIPTION
**Pull request by frolvanya**

### CODE REVIEW

**Review by olga24912 — CHANGES_REQUESTED** [review]
I think we should handle it in the following way:

(1) Check if the fee is too small, and if so, remove it immediately.
(2) Store the creation timestamp and keep it for, say, two weeks.
(3)* Skip transaction if we check the gas recently 

The problem with the transaction could be due to a high gas price on Ethereum. With an attempt-based approach, we will run out of attempts to check the gas price very quickly.

File: omni-relayer/src/startup/evm.rs [hunk 1]
```diff
@@ -99,7 +99,7 @@ async fn process_log(
             redis_connection,
             utils::redis::ETH_WITHDRAW_EVENTS,
             tx_hash.to_string(),
-            (block_number, log, tx_logs),
+            (block_number, log, tx_logs, 0),
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply omni-relayer/src/startup/evm.rs]: Fixed in https://github.com/Near-One/omni-bridge/pull/105/commits/1eb3421ed9be74ba48374119de27d40debe153a1

File: omni-relayer/src/workers/evm.rs [hunk 2]
```diff
@@ -43,9 +44,93 @@ pub async fn finalize_transfer(
...
+                                    &key,
+                                )
+                                .await;
+                            }
+
+                            continue;
+                        }
+                    }
+                    Err(err) => {
+                        warn!("Failed to check fee: {}", err);
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply omni-relayer/src/workers/evm.rs]: Fixed in https://github.com/Near-One/omni-bridge/pull/105/commits/1eb3421ed9be74ba48374119de27d40debe153a1

File: omni-relayer/src/workers/evm.rs [hunk 3]
```diff
@@ -12,11 +13,20 @@ use omni_types::{
...
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct InitTransferWithTimestamp {
+    pub block_number: u64,
+    pub log: Log,
+    pub tx_logs: Option<TransactionReceipt>,
+    pub creation_timestamp: i64,
+    pub last_update_timestamp: Option<i64>,
+}
+
 pub async fn finalize_transfer(
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply omni-relayer/src/workers/evm.rs:30]: I'll create a PR with all necessary refactors a bit later

File: omni-relayer/src/workers/evm.rs [hunk 4]
```diff
@@ -43,74 +53,165 @@ pub async fn finalize_transfer(
...
+                            )
+                            .await;
+                            return;
+                        }
+
+                        if current_timestamp
+                            - init_transfer_with_timestamp
+                                .last_update_timestamp
+                                .unwrap_or_default()
+                            < utils::redis::CHECK_INSUFFICIENT_FEE_TRANSFERS_EVERY_SECS
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply omni-relayer/src/workers/evm.rs]: Fixed in e9d04e8

File: omni-relayer/src/workers/evm.rs [hunk 5]
```diff
@@ -43,74 +53,165 @@ pub async fn finalize_transfer(
...
+
+                        let current_timestamp = chrono::Utc::now().timestamp();
+
+                        if current_timestamp - init_transfer_with_timestamp.creation_timestamp
+                            > utils::redis::KEEP_INSUFFICIENT_FEE_TRANSFERS_FOR
+                        {
                             warn!(
-                                "Failed to parse recipient as OmniAddress: {:?}",
-                                init_log.inner.recipient
+                                "Removing InitTransfer that is older than 2 weeks: {:?}",
```
[the remark this answers is not in the document]
Comment (frolvanya) [reply omni-relayer/src/workers/evm.rs]: Fixed in e9d04e8

**Review by olga24912 — APPROVED** [review]

**Pull request merged**